src/middlewared/middlewared/api/v25_04_0/filesystem.py

```diff
@@ -79,3 +86,97 @@ def payload_is_actionable(self) -> Self:
 
 class FilesystemSetPermResult(BaseModel):
     result: Literal[None]
+
+
+FILESYSTEM_STATX_ATTRS = Literal[
+    'COMPRESSED',
+    'APPEND',
+    'NODUMP',
+    'IMMUTABLE',
+    'AUTOMOUNT',
+    'MOUNT_ROOT',
+    'VERIFY',
+    'DAX'
+]
+
+
+FILESYSTEM_ZFS_ATTRS = Literal[
+    'READONLY',
+    'HIDDEN',
+    'SYSTEM',
+    'ARCHIVE',
+    'IMMUTABLE',
+    'NOUNLINK',
+    'APPENDONLY',
+    'NODUMP',
+    'OPAQUE',
+    'AV_QUARANTINED',
+    'AV_MODIFIED',
+    'REPARSE',
+    'OFFLINE',
+    'SPARSE'
+]
+
+
+class FilesystemDirEntry(BaseModel):
+    name: NonEmptyString
+    """ Entry's base name. """
+    path: NonEmptyString
+    """ Entry's full path. """
+    realpath: NonEmptyString
+    """ Canonical path of the entry, eliminating any symbolic links"""
+    type: Literal[
+        StatxEtype.DIRECTORY,
+        StatxEtype.FILE,
+        StatxEtype.SYMLINK,
+        StatxEtype.OTHER,
+    ]
+    size: int
+    """ Size in bytes of a plain file. """
+    allocation_size: int
+    mode: int
+    """ Entry's mode including file type information and file permission bits. """
+    mount_id: int
+    """ The mount ID of the mount containing the entry. This corresponds to the number in first
+    field of /proc/self/mountinfo. """
+    acl: bool
+    """ Specifies whether ACL is present on the entry. If this is the case then file permission
+    bits as reported in `mode` may not be representative of the actual permissions. """
+    uid: int
+    """ User ID of the entry's owner. """
+    gid: int
+    """ Group ID of the entry's owner. """
+    is_mountpoint: bool
+    """ Specifies whether the entry is also the mountpoint of a filesystem. """
+    is_ctldir: bool
+    """ Specifies whether the entry is located within the ZFS ctldir (for example a snapshot). """
+    attributes: list[FILESYSTEM_STATX_ATTRS]
+    """ Extra file attribute indicators for entry as returned by statx. """
+    xattrs: list[NonEmptyString]
+    """ List of xattr names of extended attributes on file. """
+    zfs_attrs: list[FILESYSTEM_ZFS_ATTRS] | None
+    """ List of extra ZFS-related file attribute indicators on file. Will be None type if filesystem is not ZFS. """
+
+
+class FilesystemListdirArgs(BaseModel):
+    path: NonEmptyString
+    query_filters: QueryFilters = []
+    query_options: QueryOptions = QueryOptions()
```

Comment:
@creatorcary I don't know if doing `QueryOptions` is the "correct" way of populating this field with default values? Do we need to do something like `Field(default_factory=QueryOptions`?

Reply:
This is how default is set in `QueryArgs` in common.py.

Reply:
Pretty sure either way works.